### PR TITLE
[WIP] Add configuration for mobile-like chat mode and recent chats list

### PR DESCRIPTION
**Pull request by raiyni**
Closes https://github.com/runelite/runelite/issues/4697 and https://github.com/runelite/runelite/issues/4833.

image

### CODE REVIEW

File: runelite-client/src/main/java/net/runelite/client/events/ChatboxInput.java [hunk 1]
```diff
@@ -29,8 +29,14 @@
 @Data
 public abstract class ChatboxInput
 {
-	private final String value;
-	private final int chatType;
+	private String value;
+	private int chatType;
+
+	public ChatboxInput(String v, int c)
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/events/ChatboxInput.java]: `@AllArgsConstructor`

File: runelite-client/src/main/java/net/runelite/client/chat/CommandManager.java [hunk 2]
```diff
@@ -125,7 +127,7 @@ private void handleInput(ScriptCallbackEvent event)
 		int stringStackCount = client.getStringStackSize();
 		int intStackCount = client.getIntStackSize();
 
-		final String typedText = stringStack[stringStackCount - 1];
+		String typedText = stringStack[stringStackCount - 1];
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/chat/CommandManager.java]: Why was this file changed?

File: runelite-client/src/main/java/net/runelite/client/plugins/clanchat/ClanChatPlugin.java [hunk 3]
```diff
@@ -25,37 +25,85 @@
...
 
 import com.google.common.eventbus.Subscribe;
+
 import java.time.temporal.ChronoUnit;
+import java.util.concurrent.ScheduledExecutorService;
 import javax.inject.Inject;
-import net.runelite.api.ChatMessageType;
-import net.runelite.api.ClanMemberRank;
+
+import com.google.inject.Provides;
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/clanchat/ClanChatPlugin.java]: Fix imports

File: runelite-client/src/main/java/net/runelite/client/chat/CommandManager.java (hunk 2 above)
Comment (raiyni) [reply runelite-client/src/main/java/net/runelite/client/chat/CommandManager.java]: Line 146. Otherwise, you can't send updated Chatbox value.

File: runelite-client/src/main/java/net/runelite/client/chat/CommandManager.java [hunk 4]
```diff
@@ -27,11 +27,13 @@
 
 import com.google.common.eventbus.EventBus;
 import com.google.common.eventbus.Subscribe;
+
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/chat/CommandManager.java]: Still, these changes are not necessary. just revert the changes in this file

File: runelite-client/src/main/java/net/runelite/client/plugins/clanchat/ClanChatPlugin.java [hunk 5]
```diff
@@ -70,6 +106,64 @@ public void updateClanChatTitle()
 		}
 	}
 
+	private boolean send(ChatboxInput chatbox)
+	{
+		executor.execute(() ->
+		{
+			chatbox.resume();
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/clanchat/ClanChatPlugin.java]: You dont need to .resume if you never called .pause

File: runelite-client/src/main/java/net/runelite/client/plugins/clanchat/ClanChatPlugin.java [hunk 6]
```diff
@@ -70,6 +106,64 @@ public void updateClanChatTitle()
...
+
+		chatboxInput.setValue(msg);
+
+		return send(chatboxInput);
+	}
+
+	@Override
+	public boolean onPrivateMessageInput(PrivateMessageInput privateMessageInput)
+	{
+		executor.execute(() ->
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/clanchat/ClanChatPlugin.java]: Same here, you just do not need to do anything in this method

File: runelite-client/src/main/java/net/runelite/client/plugins/clanchat/ClanChatPlugin.java [hunk 7]
```diff
@@ -70,6 +106,64 @@ public void updateClanChatTitle()
...
+		}
+		else
+		{
+			msg = "/" + msg;
+			chatboxInput.setChatType(2);
+		}
+
+		chatboxInput.setValue(msg);
+
+		return send(chatboxInput);
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/clanchat/ClanChatPlugin.java]: Do you really need to pause here? Will this work if you will just return false?

File: runelite-client/src/main/java/net/runelite/client/plugins/clanchat/ClanChatPlugin.java [hunk 8]
```diff
@@ -70,6 +106,64 @@ public void updateClanChatTitle()
...
+	private boolean isClanChatClosed()
+	{
+		Widget widget = client.getWidget(WidgetInfo.CLAN_CHAT);
+
+		if (widget == null)
+		{
+			return true;
+		}
+
+		return widget.isHidden();
```
Comment (Kamielvf) [inline runelite-client/src/main/java/net/runelite/client/plugins/clanchat/ClanChatPlugin.java]: Can be simplified to `return widget == null || widget.isHidden();` and removing the null check above.

File: runelite-client/src/main/java/net/runelite/client/plugins/clanchat/ClanChatPlugin.java (hunk 6 above)
Comment (raiyni) [reply runelite-client/src/main/java/net/runelite/client/plugins/clanchat/ClanChatPlugin.java]: It's abstract, have to have it.

File: runelite-client/src/main/java/net/runelite/client/plugins/clanchat/ClanChatPlugin.java (hunk 7 above)
Comment (raiyni) [reply runelite-client/src/main/java/net/runelite/client/plugins/clanchat/ClanChatPlugin.java]: That's needed, otherwise the message isn't sent to the right mode.

File: runelite-client/src/main/java/net/runelite/client/plugins/clanchat/ClanChatPlugin.java (hunk 5 above)
Comment (raiyni) [reply runelite-client/src/main/java/net/runelite/client/plugins/clanchat/ClanChatPlugin.java]: Resume sends the chat message. See, https://github.com/runelite/runelite/blob/9fe667c763e6f82d218da00a12aaf051d0defaa9/runelite-client/src/main/java/net/runelite/client/chat/CommandManager.java#L131.

It's an abstract class that is created when the event is dispatched.

File: runelite-client/src/main/java/net/runelite/client/plugins/clanchat/ClanChatPlugin.java (hunk 6 above)
Comment (deathbeam) [reply runelite-client/src/main/java/net/runelite/client/plugins/clanchat/ClanChatPlugin.java]: Yeah, but I mean the method does not need body

File: runelite-client/src/main/java/net/runelite/client/plugins/clanchat/ClanChatPlugin.java [hunk 9]
```diff
@@ -70,6 +106,50 @@ public void updateClanChatTitle()
...
+		{
+			return false;
+		}
+
+		String msg = chatboxInput.getValue();
+
+		if (msg.startsWith("/"))
+		{
+			msg = msg.substring(1);
+			chatboxInput.setChatType(0);
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/clanchat/ClanChatPlugin.java]: I believe we already have the chat types defined somewhere so maybe you should use them instead of 0/2

File: runelite-client/src/main/java/net/runelite/client/events/ChatboxInput.java [hunk 10]
```diff
@@ -25,12 +25,14 @@
 package net.runelite.client.events;
 
 import lombok.Data;
+import lombok.AllArgsConstructor;
 
 @Data
+@AllArgsConstructor
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/events/ChatboxInput.java]: Looks like you are never creating the object? So i guess this change is unnecessary too. Also I think data already has AllArgsConstructor

Comment (raiyni) [reply runelite-client/src/main/java/net/runelite/client/events/ChatboxInput.java]: Either a constructor is required or @AlllArgsConstructor is needed.

Data doesn't generate it unless the variables are final.

File: runelite-client/src/main/java/net/runelite/client/plugins/clanchat/ClanChatPlugin.java (hunk 9 above)
Comment (raiyni) [reply runelite-client/src/main/java/net/runelite/client/plugins/clanchat/ClanChatPlugin.java]: Sending chat types are different than the receiving chat types.

Comment (deathbeam) [reply runelite-client/src/main/java/net/runelite/client/plugins/clanchat/ClanChatPlugin.java]: Yeah so probably make some enum/constants for them.

**Comment by raiyni** [comment]
Closes https://github.com/runelite/runelite/issues/4833 too.

File: runelite-api/src/main/java/net/runelite/api/ChatMessageType.java [hunk 11]
```diff
@@ -48,6 +48,10 @@
 	 * A message in the public chat.
 	 */
 	PUBLIC(2),
+	/**
```
Comment (deathbeam) [inline runelite-api/src/main/java/net/runelite/api/ChatMessageType.java]: Create new enum for this, this really do not belongs here

File: runelite-client/src/main/java/net/runelite/client/plugins/clanchat/ClanChatOverlay.java [hunk 12]
```diff
@@ -0,0 +1,88 @@
...
+import net.runelite.api.GameState;
+import net.runelite.api.Point;
+import net.runelite.api.widgets.Widget;
+import net.runelite.api.widgets.WidgetInfo;
+import net.runelite.client.ui.overlay.Overlay;
+import net.runelite.client.ui.overlay.OverlayLayer;
+import net.runelite.client.ui.overlay.OverlayPosition;
+import net.runelite.client.ui.overlay.OverlayUtil;
+
+public class ClanChatOverlay extends Overlay
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/clanchat/ClanChatOverlay.java]: Remove public modifier

File: runelite-client/src/main/java/net/runelite/client/plugins/clanchat/ClanChatOverlay.java [hunk 13]
```diff
@@ -0,0 +1,88 @@
...
+	private List<String> chats = new ArrayList<>();
+
+	@Inject
+	private Client client;
+
+	@Inject
+	private ClanChatConfig config;
+
+	@Inject
+	ClanChatOverlay()
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/clanchat/ClanChatOverlay.java]: Make this constructor private and inject the dependencies in constructor param

File: runelite-client/src/main/java/net/runelite/client/plugins/clanchat/ClanChatPlugin.java [hunk 14]
```diff
@@ -70,6 +126,82 @@ public void updateClanChatTitle()
...
+		return true;
+	}
+
+	@Override
+	public boolean onPrivateMessageInput(PrivateMessageInput privateMessageInput)
+	{
+		return false;
+	}
+
+
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/clanchat/ClanChatPlugin.java]: Remove this empty line

File: runelite-client/src/main/java/net/runelite/client/plugins/clanchat/ClanChatPlugin.java [hunk 15]
```diff
@@ -70,6 +126,82 @@ public void updateClanChatTitle()
...
+	private void fixedAdd(String s)
+	{
+		if (s == null || s.isEmpty())
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/clanchat/ClanChatPlugin.java]: Strings.isNullOrEmpty

File: runelite-client/src/main/java/net/runelite/client/plugins/clanchat/ClanChatPlugin.java [hunk 16]
```diff
@@ -70,6 +126,82 @@ public void updateClanChatTitle()
...
+		{
+			return;
+		}
+
+		if (chats.stream().anyMatch(s::equalsIgnoreCase))
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/clanchat/ClanChatPlugin.java]: Why you need the if anyMatch? just do the removeIf

File: runelite-client/src/main/java/net/runelite/client/plugins/clanchat/ClanChatPlugin.java [hunk 17]
```diff
@@ -70,6 +126,82 @@ public void updateClanChatTitle()
...
+			chats.remove(0);
+		}
+
+		config.chatsData(String.join(",", chats));
+	}
+
+	@Subscribe
+	public void onVarClientstrChanged(VarClientStrChanged strChanged)
+	{
+		if (strChanged.getIndex() == 129 && config.recentChats())
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/clanchat/ClanChatPlugin.java]: Put this to VarClientStr somewhere, dont add random constants

**Comment by raiyni** [comment]
Updated preview with more padding

File: runelite-client/src/main/java/net/runelite/client/plugins/clanchat/ClanChatConfig.java [hunk 18]
```diff
@@ -0,0 +1,108 @@
...
+		return true;
+	}
+
+	@ConfigItem(
+		keyName = "recentColor",
+		name = "Recent Chats Color",
+		description = "Color configuration for recent chats.",
+		position = 5
+	)
+	default Color recentColor()
```
[the remark this answers is not in the document]
Comment (raiyni) [reply runelite-client/src/main/java/net/runelite/client/plugins/clanchat/ClanChatConfig.java]: I'll have to check on that.

**Comment by raiyni** [comment]
Okay, I didn't go through with making it a widget. Left it as an overlay due to my struggles with dynamically creating widgets. I'll detail the issue if someone else knows more about it.

```
Widget widget = client.getWidget(WidgetInfo.CLAN_CHAT_LIST);

Widget child = widget.createChild(0, 4);
child.setText("TEXT");
child.setColor(0xffffff);
child.setWidth(154);
child.setHeight(15);
```

creates and adds a child to the clan chat list but the text isn't draw. If you add an action to it

```
child.setAction(1, "TEST");
child.setScript(ScriptID.NULL);
```

you get the TEST menu when you mouse over it.

If you change it to a type 5 and set a spriteid

```
child.setType(5);
child.setSpriteId(SpriteID.GE_SEARCH);
```

it will display the icon.

Did this at onGameTick and onRender. It does the same thing.

File: runelite-client/src/main/java/net/runelite/client/plugins/clanchat/ClanChatPlugin.java [hunk 19]
```diff
@@ -25,37 +25,97 @@
...
+
+	@Inject
+	private ClanChatOverlay overlay;
+
+	@Inject
+	private OverlayManager overlayManager;
+
+	private List<String> chats = new ArrayList<>();
+
+	private final int MAX_CHATS = 10;
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/clanchat/ClanChatPlugin.java]: private static final, and put it near beginning of the class

File: runelite-client/src/main/java/net/runelite/client/plugins/clanchat/ClanChatPlugin.java [hunk 20]
```diff
@@ -25,37 +25,97 @@
...
+
 	@Schedule(
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/clanchat/ClanChatPlugin.java]: What is this leftover? Looks like it should be above updateClanChatTitle. Also, maybe changing that to be suscribed to GameTIck would be better instead (and naming the method onGameTick)

File: runelite-client/src/main/java/net/runelite/client/plugins/clanchat/ClanChatPlugin.java [hunk 21]
```diff
@@ -70,6 +130,79 @@ public void updateClanChatTitle()
...
+		executor.execute(() ->
+		{
+			chatboxInput.resume();
+		});
+
+		return true;
+	}
+
+	@Override
+	public boolean onPrivateMessageInput(PrivateMessageInput privateMessageInput)
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/clanchat/ClanChatPlugin.java]: You should usually put overrides together (e.g near the startup/shutdown)

**Comment by deathbeam** [comment]
Also, instead of adding Recent chats at top, I would change the CC title to say "Clan Chat - Last joined" or something

File: runelite-client/src/main/java/net/runelite/client/plugins/clanchat/ClanChatPlugin.java (hunk 20 above)
Comment (raiyni) [reply runelite-client/src/main/java/net/runelite/client/plugins/clanchat/ClanChatPlugin.java]: I don't know, it was already there

File: runelite-client/src/main/java/net/runelite/client/plugins/clanchat/ClanChatOverlay.java [hunk 22]
```diff
@@ -0,0 +1,94 @@
...
+import net.runelite.client.ui.overlay.OverlayPosition;
+import net.runelite.client.ui.overlay.OverlayUtil;
+
+class ClanChatOverlay extends Overlay
+{
+	@Getter
+	@Setter
+	private List<String> chats = new ArrayList<>();
+
+	private Client client;
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/clanchat/ClanChatOverlay.java]: final on both client and config

File: runelite-client/src/main/java/net/runelite/client/plugins/clanchat/ClanChatOverlay.java [hunk 23]
```diff
@@ -0,0 +1,94 @@
...
+import net.runelite.api.widgets.WidgetInfo;
+import net.runelite.client.ui.FontManager;
+import net.runelite.client.ui.overlay.Overlay;
+import net.runelite.client.ui.overlay.OverlayLayer;
+import net.runelite.client.ui.overlay.OverlayPosition;
+import net.runelite.client.ui.overlay.OverlayUtil;
+
+class ClanChatOverlay extends Overlay
+{
+	@Getter
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/clanchat/ClanChatOverlay.java]: The getter seems unused

File: runelite-client/src/main/java/net/runelite/client/plugins/clanchat/ClanChatOverlay.java [hunk 24]
```diff
@@ -0,0 +1,94 @@
...
+	public Dimension render(Graphics2D graphics)
+	{
+		if ((client.getGameState() != GameState.LOADING && client.getGameState() != GameState.LOGGED_IN) || !config.recentChats())
+		{
+			return null;
+		}
+
+		Widget widget = client.getWidget(WidgetInfo.CLAN_CHAT.getGroupId(), 15);
+
+
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/clanchat/ClanChatOverlay.java]: remove this empty line

File: runelite-client/src/main/java/net/runelite/client/plugins/clanchat/ClanChatPlugin.java [hunk 25]
```diff
@@ -25,38 +25,124 @@
...
+	@Inject
+	private ClanChatConfig config;
+
+	@Inject
+	private ClanChatOverlay overlay;
+
+	@Inject
+	private OverlayManager overlayManager;
+
+	private List<String> chats = new ArrayList<>();
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/clanchat/ClanChatPlugin.java]: Looks like you dont need this property as you are sending it directly to overlay

Comment (raiyni) [reply runelite-client/src/main/java/net/runelite/client/plugins/clanchat/ClanChatPlugin.java]: Need it for adding to the list. Gets passed by reference so it is set on the overlay once and can be added/popped on the plugin.

File: runelite-client/src/main/java/net/runelite/client/plugins/clanchat/ClanChatOverlay.java [hunk 26]
```diff
@@ -0,0 +1,91 @@
...
+		}
+
+		Widget widget = client.getWidget(WidgetInfo.CLAN_CHAT.getGroupId(), 15);
+
+		if (widget != null && !widget.isHidden() && widget.getChildren() == null)
+		{
+			Point location = widget.getCanvasLocation();
+
+			graphics.setFont(font);
+			OverlayUtil.setGraphicProperties(graphics);
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/clanchat/ClanChatOverlay.java]: You dont need to do this. Its already being set from the OVerlayManager. Just setting the font on graphics is enough

File: runelite-client/src/main/java/net/runelite/client/plugins/clanchat/ClanChatConfig.java [hunk 27]
```diff
@@ -0,0 +1,95 @@
...
+
+import net.runelite.client.config.Config;
+import net.runelite.client.config.ConfigGroup;
+import net.runelite.client.config.ConfigItem;
+
+@ConfigGroup("clanchat")
+public interface ClanChatConfig extends Config
+{
+	@ConfigItem(
+		keyName = "rankIcons",
```
[the remark this answers is not in the document]
Comment (raiyni) [reply runelite-client/src/main/java/net/runelite/client/plugins/clanchat/ClanChatConfig.java]: it's set to that

[the remark this answers is not in the document]
Comment (raiyni) [reply runelite-client/src/main/java/net/runelite/client/plugins/clanchat/ClanChatConfig.java]: github defaults to 8 if memory serves for viewing files

File: runelite-client/src/main/java/net/runelite/client/plugins/clanchat/ClanChatPlugin.java (hunk 14 above)
Comment (deathbeam) [reply runelite-client/src/main/java/net/runelite/client/plugins/clanchat/ClanChatPlugin.java]: This empty line is still here

Comment (raiyni) [reply runelite-client/src/main/java/net/runelite/client/plugins/clanchat/ClanChatPlugin.java]: I removed it again. And again.

Comment (raiyni) [reply runelite-client/src/main/java/net/runelite/client/plugins/clanchat/ClanChatPlugin.java]: if you click View for the file, it's gone
